### PR TITLE
Events

### PR DESCRIPTION
**Pull request by smizell**
Note: this is experimental at this point. Wanting to see what it would look like to be able to listen to changes at different nodes in the tree.

### CODE REVIEW

File: lib/primitives/array-element.js [hunk 1]
```diff
@@ -93,7 +96,12 @@ module.exports = function(BaseElement, registry) {
     },
 
     set: function(index, value) {
-      this.content[index] = registry.toElement(value);
+      var item = registry.toElement(value);
+      if (!_.contains(item.parents, this)) {
+        item.parents.push(this);
+      }
+      this.content[index] = item;
+      this.emitSwap();
```
Comment (fosrias) [inline lib/primitives/array-element.js:104]: swap? Sort of confusing semantics. This is replacing the element at some position with another, is that correct?

File: lib/primitives/base-element.js [hunk 2]
```diff
@@ -132,6 +135,19 @@ module.exports = function(registry) {
 
     setMetaProperty: function(name, value) {
       this.meta.set(name, value);
+    },
+
+    emitSwap: function() {
```
Comment (fosrias) [inline lib/primitives/base-element.js:140]: This does not seem to correlate the objects related to a particular event as to what was replaced by what if I understand this correctly. Seems like something is missing with respect to the event. Should this have some event object that contains the information?

File: test/primitives/array-element-test.js [hunk 3]
```diff
@@ -128,6 +128,26 @@ describe('ArrayElement', function() {
...
+      });
+
+      it('triggers swap on the parents', function(done) {
+        var arr = new minim.ArrayElement([1]);
+        arr.on('swap', done);
+        arr.set(1, 2);
+      });
+
+      context('when a value is changed directly', function() {
+        it('triggers swap on the parents', function(done) {
```
Comment (fosrias) [inline test/primitives/array-element-test.js:145]: Is this "bubbling" the event, or actually a separate event?

File: test/primitives/array-element-test.js [hunk 4]
```diff
@@ -176,6 +196,18 @@ describe('ArrayElement', function() {
         arrayElement.push('foobar');
         itAddsToArray(arrayElement);
       });
+
+      it('sets the parent to children', function() {
+        var item = new minim.StringElement('foobar');
+        arrayElement.push(item);
+        expect(item.parents).to.include(arrayElement);
```
Comment (fosrias) [inline test/primitives/array-element-test.js:203]: Seems this functionality should be private and associated only with event functionality if that is really the purpose of parents here. Not clear.

File: test/primitives/object-element-test.js [hunk 5]
```diff
@@ -207,6 +207,47 @@ describe('ObjectElement', function() {
       obj.set({ foo: 'bar' });
       expect(obj.get('foo').toValue()).to.equal('bar');
     });
+
+    it('sets the parent to children', function() {
```
Comment (fosrias) [inline test/primitives/object-element-test.js:211]: Are we normally tracking parents or why are we tracking parents?

File: lib/primitives/array-element.js (hunk 1 above)
Comment (smizell) [inline lib/primitives/array-element.js:104]: Swap is a term used for when new instances and memory allocations are changed in atomic structures. Swap means that a node on the tree has changed. See description here http://omniscientjs.github.io/. 

Note: I want to add more events that can be listened to, such as `update` or `add`.

File: lib/primitives/base-element.js (hunk 2 above)
Comment (smizell) [inline lib/primitives/base-element.js:140]: I would handle passing the changed value in another event, like `update`. This is merely saying the branch of the tree has changed.

File: test/primitives/array-element-test.js (hunk 3 above)
Comment (smizell) [inline test/primitives/array-element-test.js:145]: Each event is on a separate node. So when you trigger an event on a node, it triggers the swap events all of the way up. This means you can listen for changes on specific parts of the tree.

File: test/primitives/object-element-test.js (hunk 5 above)
Comment (smizell) [inline test/primitives/object-element-test.js:211]: Parents let you go up and down the tree. Right now, if you get a node from the tree, you cannot know where it is in the tree, only what its children are.

File: test/primitives/array-element-test.js (hunk 4 above)
Comment (smizell) [inline test/primitives/array-element-test.js:203]: Parent just provides access to the node's parents. This is useful in functions like `find` where you want to find certain elements that have a specific kind of parent. Also allows traversing up the tree.

**Comment by kylef** [comment]
Closing, this was solved by #141.

**Pull request closed without merging**